### ISSUE TITLE
compatibility with php7 ?

### ISSUE THREAD

**Issue by pesseyjulien** [doctrine/DoctrineMongoDBBundle#351]
Is the support of php7, and the new mongodb driver, planned ?

**Comment by stof** [comment]
Well, this should be asked on the ODM first.

Btw, there is already a way to run it on PHP 7 thanks to an adapter shimming ext-mongo on top of ext-mongodb: http://www.doctrine-project.org/2016/02/16/doctrine-mongodb-odm-release-1.0.5.html

**Comment by malarzm** [comment]
@pesseyjulien yes it is planned, you can subscribe to https://github.com/doctrine/mongodb-odm/issues/1292 where updates are posted. For now the only way is one in described in blogpost @stof linked to

**Comment by pesseyjulien** [comment]
Thanks for the quick answers, and sorry if I did not post this question in the right place.

I have followed @stof link but I'm getting the following error :

 An error occurred when executing the "'cache:clear --no-warmup'" command:    
  Fatal error: Cannot use 'String' as class name as it is reserved in /Users/  
  pessey/Dropbox/Codes/web/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/Mongo  
  DB/Mapping/Annotations/String.php on line 26  

Is that related to you or doctrine/mongodb-odm ?

Thanks again,
julien

**Comment by malarzm** [comment]
@pesseyjulien please check your code first as ODM doesn't auto import these classes for PHP 7 - https://github.com/doctrine/mongodb-odm/blob/61f8e07380e41dce6ff774477ddc9da9768c43e8/lib/Doctrine/ODM/MongoDB/Mapping/Annotations/DoctrineAnnotations.php#L42

**Comment by pesseyjulien** [comment]
Thanks again for the quick answer. I have checked among my files and this is the output :


```
/Users/pessey/Dropbox/Codes/web/vendor/alcaeus/mongo-php-adapter/lib/Mongo/MongoClient.php:
  314       * String representation of this connection
  315       *
  316:      * @link http://www.php.net/manual/en/mongoclient.tostring.php
  317       * @return string Returns hostname and port for this connection.
  318       */

/Users/pessey/Dropbox/Codes/web/vendor/alcaeus/mongo-php-adapter/lib/Mongo/MongoCollection.php:
   80       * String representation of this collection
   81       *
   82:      * @link http://www.php.net/manual/en/mongocollection.--tostring.php
   83       * @return string Returns the full name of this collection.
   84       */

/Users/pessey/Dropbox/Codes/web/vendor/alcaeus/mongo-php-adapter/lib/Mongo/MongoDB.php:
   81       * The name of this database
   82       *
   83:      * @link http://www.php.net/manual/en/mongodb.--tostring.php
   84       * @return string Returns this database's name.
   85       */

/Users/pessey/Dropbox/Codes/web/vendor/alcaeus/mongo-php-adapter/lib/Mongo/MongoId.php:
   58      /**
   59       * Returns a hexidecimal representation of this id
   60:      * @link http://www.php.net/manual/en/mongoid.tostring.php
   61       * @return string
   62       */

/Users/pessey/Dropbox/Codes/web/vendor/doctrine/mongodb/lib/Doctrine/MongoDB/Collection.php:
  871       * Wrapper method for MongoCollection::__toString().
  872       *
  873:      * @see http://www.php.net/manual/en/mongocollection.--tostring.php
  874       * @return string
  875       */

/Users/pessey/Dropbox/Codes/web/vendor/doctrine/mongodb/lib/Doctrine/MongoDB/Connection.php:
  390       * Wrapper method for MongoClient::__toString().
  391       *
  392:      * @see http://php.net/manual/en/mongoclient.tostring.php
  393       * @return string
  394       */

/Users/pessey/Dropbox/Codes/web/vendor/doctrine/mongodb/lib/Doctrine/MongoDB/Database.php:
  512       * Wrapper method for MongoDB::__toString().
  513       *
  514:      * @see http://www.php.net/manual/en/mongodb.--tostring.php
  515       * @return string
  516       */

/Users/pessey/Dropbox/Codes/web/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/Mapping/Annotations/DoctrineAnnotations.php:
   44      require_once __DIR__ . '/Int.php';
   45      require_once __DIR__ . '/Float.php';
   46:     require_once __DIR__ . '/String.php';
   47  }
   48  

/Users/pessey/Dropbox/Codes/web/vendor/guzzle/guzzle/docs/webservice-client/guzzle-service-descriptions.rst:
   39          "operations": {},
   40          "models": {},
   41:         "includes": ["string.php", "string.json"]
   42      }
   43  
```

No trace of String.php except in DoctrineAnnotations.php, but it should not indeed load it as my PHP_VERSION_ID = 70000.

Do you know if I have something missing ?

**Comment by malarzm** [comment]
Hmmm are you sure you have no document fields mapped with `@String`? Otherwise you may try warming the cache with `-vvv` and maybe stack trace will tell us more

**Comment by pesseyjulien** [comment]
Hum, in that case I think I do indeed ! I was looking for the use of the class String.php itself, not the annotations. So you are saying that  `@MongoDB\String` is not valid anymore ? If not, what should I use instead ?

**Comment by alcaeus** [comment]
These annotations are deprecated and will be removed in 2.0, partly because of this incompatibility with PHP 7. You'll have to change the mapping to use the `@Field` annotation:
```
/**
 * @MongoDB\Field(type="string")
 */
protected $myProperty;
```

Same goes for other types (integer/int, boolean/bool and float) - these must be changed to be able to run on PHP 7. The deprecation affects many more annotations.

**Comment by pesseyjulien** [comment]
Alright, thanks a bunch ! I will update my code then.

**Comment by webdevilopers** [comment]
Hey @alcaeus ,
I've been fighting with this issue quite a while now.

At first I thought it would be caused by the Gedmo Doctrine Extensions:
* https://github.com/stof/StofDoctrineExtensionsBundle/issues/332
* https://github.com/Atlantic18/DoctrineExtensions/pull/1619

No I removed all these bundles and dependencies. I removed any import of the annotation. Replaced every existing `@String` as you suggested.
My IDE will not find any usage.

I checked the PHP version line:
```php
// Don't import annotations whose names are reserved words in PHP7+
if (PHP_VERSION_ID < 70000) {
    require_once __DIR__ . '/Bool.php';
    require_once __DIR__ . '/Int.php';
    require_once __DIR__ . '/Float.php';
    require_once __DIR__ . '/String.php';
}
```

It's not called. I cleared Doctrine Cache. I tried removing the Symfony Cache even manually.
Whenever I do a warmup or clear I still get:
**Fatal error: Cannot use 'String' as class name as it is reserved in /var/www/example.com/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/Mapping/Annotations/String.php on line 26**

Any idea what I could be missing? All dependencies run on `dev-master`.
